File: frontend-react/project/src/features/processing/processingEnums.ts
/**
 * Central mapping for the backend's Processing enums.
 *
 * The API returns enums as strings ("Sorting") but the backend does NOT register a
 * JsonStringEnumConverter, so enums inside REQUEST BODIES must be sent as numbers.
 * The numeric values below mirror the C# declaration order in
 * backend/.../Features/Processing/Entities/Enums.cs — append-only, never reorder.
 * Nothing outside this file should contain a raw enum number.
 */

// ---------------------------------------------------------------- enum value maps

export const INVENTORY_STATUS_VALUES = {
  Received: 0,
  Sorting: 1,
  Dismantling: 2,
  Classified: 3,
  ReadyForSale: 4,
  ExportOnly: 5,
  OnHold: 6,
} as const;

export const CLASSIFICATION_CATEGORY_VALUES = {
  Reusable: 0,
  LocalRecyclable: 1,
  Hazardous: 2,
  ExportOnly: 3,
} as const;

export const CLASSIFICATION_SOURCE_VALUES = {
  Manual: 0,
  Ai: 1,
} as const;

export const ORIGIN_TYPE_VALUES = {
  JobCollection: 0,
  ExtraWaste: 1,
} as const;

export const PAYMENT_SOURCE_TYPE_VALUES = {
  Job: 0,
  ExtraWaste: 1,
} as const;

export const PAYMENT_STATUS_VALUES = {
  Pending: 0,
  Paid: 1,
} as const;

// ---------------------------------------------------------------- string-name types

export type InventoryStatus = keyof typeof INVENTORY_STATUS_VALUES;
export type ClassificationCategory = keyof typeof CLASSIFICATION_CATEGORY_VALUES;
export type ClassificationSource = keyof typeof CLASSIFICATION_SOURCE_VALUES;
export type OriginType = keyof typeof ORIGIN_TYPE_VALUES;
export type PaymentSourceType = keyof typeof PAYMENT_SOURCE_TYPE_VALUES;
export type PaymentStatus = keyof typeof PAYMENT_STATUS_VALUES;

export const INVENTORY_STATUSES = Object.keys(INVENTORY_STATUS_VALUES) as InventoryStatus[];
export const CLASSIFICATION_CATEGORIES = Object.keys(CLASSIFICATION_CATEGORY_VALUES) as ClassificationCategory[];
export const CLASSIFICATION_SOURCES = Object.keys(CLASSIFICATION_SOURCE_VALUES) as ClassificationSource[];
export const ORIGIN_TYPES = Object.keys(ORIGIN_TYPE_VALUES) as OriginType[];
export const PAYMENT_SOURCE_TYPES = Object.keys(PAYMENT_SOURCE_TYPE_VALUES) as PaymentSourceType[];
export const PAYMENT_STATUSES = Object.keys(PAYMENT_STATUS_VALUES) as PaymentStatus[];

// ---------------------------------------------------------------- name -> number (request bodies)

export const inventoryStatusValue = (s: InventoryStatus): number => INVENTORY_STATUS_VALUES[s];
export const classificationCategoryValue = (c: ClassificationCategory): number => CLASSIFICATION_CATEGORY_VALUES[c];
export const classificationSourceValue = (s: ClassificationSource): number => CLASSIFICATION_SOURCE_VALUES[s];

// ---------------------------------------------------------------- type guards (URL params etc.)

const isOneOf = <T extends string>(list: readonly T[], value: unknown): value is T =>
  typeof value === 'string' && (list as readonly string[]).includes(value);

export const isInventoryStatus = (v: unknown): v is InventoryStatus => isOneOf(INVENTORY_STATUSES, v);
export const isClassificationCategory = (v: unknown): v is ClassificationCategory => isOneOf(CLASSIFICATION_CATEGORIES, v);
export const isOriginType = (v: unknown): v is OriginType => isOneOf(ORIGIN_TYPES, v);
export const isPaymentSourceType = (v: unknown): v is PaymentSourceType => isOneOf(PAYMENT_SOURCE_TYPES, v);
export const isPaymentStatus = (v: unknown): v is PaymentStatus => isOneOf(PAYMENT_STATUSES, v);

// ---------------------------------------------------------------- display labels

export const INVENTORY_STATUS_LABELS: Record<InventoryStatus, string> = {
  Received: 'Received',
  Sorting: 'Sorting',
  Dismantling: 'Dismantling',
  Classified: 'Classified',
  ReadyForSale: 'Ready for sale',
  ExportOnly: 'Export only',
  OnHold: 'On hold',
};

export const CLASSIFICATION_CATEGORY_LABELS: Record<ClassificationCategory, string> = {
  Reusable: 'Reusable',
  LocalRecyclable: 'Local recyclable',
  Hazardous: 'Hazardous',
  ExportOnly: 'Export only',
};

export const CLASSIFICATION_SOURCE_LABELS: Record<ClassificationSource, string> = {
  Manual: 'Manual',
  Ai: 'AI-assisted',
};

export const ORIGIN_TYPE_LABELS: Record<OriginType, string> = {
  JobCollection: 'Job collection',
  ExtraWaste: 'Extra waste',
};

export const PAYMENT_SOURCE_TYPE_LABELS: Record<PaymentSourceType, string> = {
  Job: 'Job collection',
  ExtraWaste: 'Extra waste',
};

export const PAYMENT_STATUS_LABELS: Record<PaymentStatus, string> = {
  Pending: 'Pending',
  Paid: 'Paid',
};

// ---------------------------------------------------------------- status workflow

/**
 * Mirror of InventoryItem.AllowedTransitions on the backend. The server stays the source of
 * truth (it answers 409 on an illegal jump); this copy only lets the UI hide actions that
 * could never succeed.
 */
export const ALLOWED_TRANSITIONS: Record<InventoryStatus, readonly InventoryStatus[]> = {
  Received: ['Sorting'],
  Sorting: ['Dismantling', 'Classified'],
  Dismantling: ['Classified'],
  Classified: ['ReadyForSale', 'ExportOnly', 'OnHold'],
  ReadyForSale: [],
  ExportOnly: [],
  OnHold: [],
};

export const isTerminalStatus = (s: InventoryStatus): boolean => ALLOWED_TRANSITIONS[s].length === 0;

/** Dismantle steps are only accepted while an item is being sorted or dismantled. */
export const canAddDismantleStep = (s: InventoryStatus): boolean => s === 'Sorting' || s === 'Dismantling';

/** Classification is only accepted from Sorting/Dismantling, and is the only way to reach Classified. */
export const canClassify = (s: InventoryStatus): boolean => s === 'Sorting' || s === 'Dismantling';

/**
 * Transitions the plain status endpoint may perform. Classified is reachable only through
 * PUT /classify (the backend rejects it here), and Dismantling is entered by logging the first
 * dismantle step, so both are excluded.
 */
export const getManualTransitions = (s: InventoryStatus): InventoryStatus[] =>
  ALLOWED_TRANSITIONS[s].filter((next) => next !== 'Classified' && next !== 'Dismantling');

/**
 * "GeneralCollection" is the rate-policy key that prices the weight part of a JOB-collection
 * payment. It is not a real item type, so it is hidden from extra-waste item selection (the backend
 * rejects it too). It still appears on the read-only Rate Policies page.
 */
export const JOB_PAYMENT_RATE_KEY = 'GeneralCollection';

export const isReservedExtraWasteType = (itemType: string): boolean =>
  itemType.trim().toLowerCase() === JOB_PAYMENT_RATE_KEY.toLowerCase();

/** Backend limits, kept next to the enums so forms and validators agree with the API. */
export const LIMITS = {
  notes: 1000,
  subCategory: 100,
  childItemType: 50,
  itemType: 50,
  rejectionReason: 500,
  idempotencyKey: 100,
  search: 100,
  pageSize: 20,
  maxPageSize: 100,
} as const;

/** ClassificationValidationService flags anything below this for human review. */
export const LOW_CONFIDENCE_THRESHOLD = 0.7;

File: mobile_flutter/lib/features/warehouse/data/processing_enums.dart
/// The backend's Processing enums — a port of the web app's processingEnums.ts.
///
/// The API RETURNS enums as strings ("Sorting") but only ACCEPTS numbers in request bodies (no
/// JsonStringEnumConverter is registered), so each value keeps its C# declaration order as
/// [index]. Append-only, never reorder — it must match
/// backend/.../Features/Processing/Entities/Enums.cs.
library;

enum InventoryStatus {
  received('Received', 'Received'),
  sorting('Sorting', 'Sorting'),
  dismantling('Dismantling', 'Dismantling'),
  classified('Classified', 'Classified'),
  readyForSale('ReadyForSale', 'Ready for sale'),
  // Named differently from the "Export-grade" category so the two are never confused.
  exportOnly('ExportOnly', 'Reserved for export'),
  onHold('OnHold', 'On hold');

  const InventoryStatus(this.apiName, this.label);

  final String apiName;
  final String label;

  /// The number to send in a request body.
  int get apiValue => index;

  static InventoryStatus fromApi(String name) =>
      values.firstWhere((s) => s.apiName == name, orElse: () => throw FormatException('Unknown status $name'));

  /// Mirror of InventoryItem.AllowedTransitions. The server stays the source of truth; this only
  /// hides actions that could never succeed.
  List<InventoryStatus> get allowedNext => switch (this) {
        received => const [sorting],
        sorting => const [dismantling, classified],
        dismantling => const [classified],
        classified => const [readyForSale, exportOnly, onHold],
        readyForSale || exportOnly || onHold => const [],
      };

  bool get isTerminal => allowedNext.isEmpty;

  /// Dismantle steps are only accepted while an item is being sorted or dismantled.
  bool get canDismantle => this == sorting || this == dismantling;

  /// Classifying is only accepted from Sorting/Dismantling, and is the only way to reach Classified.
  bool get canClassify => this == sorting || this == dismantling;

  /// Transitions the plain status endpoint may perform: Classified is reached only through
  /// /classify and Dismantling by logging the first dismantle step. From Classified, only the
  /// category's outcome and On hold (ClassificationOutcomeRules on the backend).
  List<InventoryStatus> manualTransitions(ClassificationCategory? category) => allowedNext
      .where((next) => next != classified && next != dismantling)
      .where((next) => this != classified || next == onHold || category?.outcome == next)
      .toList();

  /// The warehouse area this status normally lives in (seeded location names). Only used to
  /// pre-select the location when changing status; staff can always pick another.
  String? get suggestedLocationName => switch (this) {
        sorting => 'Sorting Area',
        readyForSale => 'Ready-for-Sale Storage',
        exportOnly => 'Export Storage',
        onHold => 'Hazardous Hold Area',
        _ => null,
      };
}

enum ClassificationCategory {
  reusable('Reusable', 'Reusable', 'Can be refurbished or reused as it is. Next step: ready for sale.'),
  localRecyclable('LocalRecyclable', 'Local recyclable', 'Can be recycled through local partners. Next step: ready for sale.'),
  hazardous('Hazardous', 'Hazardous',
      'Contains hazardous material (batteries, CRT glass, mercury, lead…). Quarantined automatically.'),
  exportOnly('ExportOnly', 'Export-grade', 'Can only be handled through export channels. Next step: reserved for export.');

  const ClassificationCategory(this.apiName, this.label, this.help);

  final String apiName;
  final String label;
  final String help;

  int get apiValue => index;

  /// The outcome the category leads to; Hazardous has none (it goes on hold automatically).
  InventoryStatus? get outcome => switch (this) {
        reusable || localRecyclable => InventoryStatus.readyForSale,
        exportOnly => InventoryStatus.exportOnly,
        hazardous => null,
      };

  static ClassificationCategory? tryFromApi(String? name) =>
      values.where((c) => c.apiName == name).firstOrNull;
}

enum ClassificationSource {
  manual('Manual', 'Manual'),
  ai('Ai', 'AI-assisted');

  const ClassificationSource(this.apiName, this.label);

  final String apiName;
  final String label;

  int get apiValue => index;

  static ClassificationSource? tryFromApi(String? name) => values.where((s) => s.apiName == name).firstOrNull;
}

enum OriginType {
  jobCollection('JobCollection', 'Job collection'),
  extraWaste('ExtraWaste', 'Extra waste');

  const OriginType(this.apiName, this.label);

  final String apiName;
  final String label;

  static OriginType? tryFromApi(String? name) => values.where((o) => o.apiName == name).firstOrNull;
}

/// "GeneralCollection" is the rate that prices the weight part of a job payment. It is not a real
/// item type, so it is never offered as an extra-waste item (the backend rejects it too).
const jobPaymentRateKey = 'GeneralCollection';

bool isReservedItemType(String itemType) => itemType.trim().toLowerCase() == jobPaymentRateKey.toLowerCase();

/// Backend limits, so forms agree with the API validators.
abstract final class Limits {
  static const notes = 1000;
  static const subCategory = 100;
  static const itemType = 50;
  static const rejectionReason = 500;
  static const search = 100;
  static const pageSize = 20;
}

/// ClassificationValidationService flags anything below this for human review.
const lowConfidenceThreshold = 0.7;

/// What an inventory QR label holds: "EWI:" + the inventory item id. The prefix lets the scanner
/// ignore any other QR code it sees. A bare id is accepted too (e.g. typed in by hand).
const inventoryQrPrefix = 'EWI:';

String inventoryQrData(String itemId) => '$inventoryQrPrefix$itemId';

final _guid = RegExp(r'^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$', caseSensitive: false);

/// The inventory item id in a scanned/typed value, or null if it isn't one of our labels.
String? parseInventoryQr(String? raw) {
  if (raw == null) return null;
  var value = raw.trim();
  if (value.toUpperCase().startsWith(inventoryQrPrefix)) value = value.substring(inventoryQrPrefix.length).trim();
  return _guid.hasMatch(value) ? value.toLowerCase() : null;
}
